### PR TITLE
Optimizations

### PR DESCRIPTION
**Pull request by sebasv**
I rewrote the nelder-mead code to the following ends:
* moved the 2d array into a vec of 1d arrays. This way, a sort only needs to move pointers instead of the actual arrays.
* changed the sorting such that sorting an array that was sorted, but had 1 element inserted, takes O(n) steps and (barely) no mem allocs, instead of O(n*n) allocs
* changed the centroid update code to be O(n) instead of O(n*n)
* added a reference

@to266 I used the same reference as the one used in SciPy,. If you have a more appropriate reference, I can update the docstring!

### CODE REVIEW

File: src/vector/nelder_mead.rs [hunk 1]
```diff
@@ -60,178 +85,176 @@ pub struct NelderMead {
...
+        (maxiter, maxfun, alpha, beta, gamma, delta)
+    }
 
-            let mut doshrink = false;
+    #[inline]
+    fn finished(&self, simplex: &Simplex, iterations: usize, maxiter: usize, nfeval: usize, maxfun: usize) -> bool {
+        let n = simplex.len();
+        if iterations > maxiter || nfeval > maxfun {
+            true
+        } else if simplex[n-1].0 - simplex[0].0 < self.ftol
```
[the remark this answers is not in the document]
Comment (sebasv) [reply src/vector/nelder_mead.rs]: changed

**Comment by sebasv** [comment]
Come to think of it, `minimize_x0` was something that seemed natural when I wrote the `minimize_simplex` routine but not having a default `minimize` seems like a loss.

**Comment by to266** [comment]
Excellent work! If you're going to keep this pace, do create issues / pre-emptive PRs just to coordinate so that we don't bump into each other, since I expect to find more time for this crate soon

**Comment by sebasv** [comment]
Allright, I will set up preemptive pull requests as needed!

Met vriendelijke groet, with kind regards,

Sebastiaan Vermeulen
---------------------
PhD Candidate -- Erasmus School of Economics

T +31 10 408 8924
E vermeulen@ese.eur.nl
W https://sebastiaanvermeulen.nl

________________________________
Van: Tomas Ostasevicius <notifications@github.com>
Verzonden: Thursday, October 11, 2018 8:37:07 AM
Aan: to266/optimize
CC: Sebastiaan Vermeulen; Author
Onderwerp: Re: [to266/optimize] Optimizations (#5)


Excellent work! If you're going to keep this pace, do create issues / pre-emptive PRs just to coordinate so that we don't bump into each other, since I expect to find more time for this crate soon

**Pull request merged**